AI/Building LLM system/graphrag.md

```diff
@@ -0,0 +1,73 @@
+---
+tags:
+  - llm
+  - rag
+title: GraphRAG - Building a knowledge graph for RAG system
+description: "In baseline Retrieval Augmented Generation (RAG), sometimes the result might not be accurate as expected since the query itself have multiple layers of reasoning or the answer requires traversing disparate pieces of information through their shared attributes in order to provide new synthesized insights. In this post, we will explore a new approach called GraphRAG which combines the strengths of knowledge graphs and large language models to improve the accuracy of RAG systems"
+date: 2024-11-01
+authors:
+  - hoangnnh
+---
+
+In baseline Retrieval Augmented Generation (RAG), sometimes the result might not be accurate as expected since the query itself have multiple layers of reasoning or the answer requires traversing disparate pieces of information through their shared attributes in order to provide new synthesized insights. In this post, we will explore a new approach called GraphRAG which combines the strengths of knowledge graphs and large language models to improve the accuracy of RAG systems.
+
+## What is Knowledge Graph?
+
+A knowledge graph is an organized representation of real-world entities and their relationships. It is typically stored in a graph database, which natively stores the relationships between data entities. Entities in a knowledge graph can represent objects, events, situations, or concepts. Knowledge graphs contain 2 key chracteristics: 
+ - **Nodes**: Represent entities such as people, places, organizations, events, or concepts,... Each node can have properties or attributes that describe it. For example, A node with type Person might have properties like name, age, and occupation.
+ - **Edges**: Represent the relationships or connections between entities. Edges can have types and properties as well. For example, an edge with type FRIEND_OF might have a property called "since", indicating when the friendship began.
+
+![Knowledge Graph](assets/graphrag-knowledge-graph.webp)
+
+## Why Knowledge Graph is used in RAG?
+
+Naive RAG systems built with keyword or similarity search-based retrieval fail in complex queries that require reasoning. Suppose user asks a query: "Who directed the sci-fi movie where the lead actor was also in The Revenant?", a standard RAG system will retrieve documents about The Revenant, then find information about cast and crew members, but it will fail identiy the lead actor, Leo DiCaprio who starred in other movies but not existed in the query to retrieve document. However, taking advantage of knowledge graph, the ideally process will be: identify the lead actor of The Revenant, then find other movies he starred in, then find the sci-fi movie which meet ciriteria.
```

Comment:
```suggestion
Naive RAG systems built with keyword or similarity search-based retrieval fail in complex queries that require reasoning. Suppose user asks a query: "Who directed the sci-fi movie where the lead actor was also in The Revenant?", a standard RAG system will retrieve documents about The Revenant, then find information about cast and crew members, but it will fail to identify the lead actor, Leo DiCaprio who starred in other movies but not existed in the query to retrieve document. However, taking advantage of knowledge graph, the ideally process will be: identify the lead actor of The Revenant, then find other movies he starred in, then find the sci-fi movie which meet criteria.
```